libs/cardano-ledger-conformance/src/Test/Cardano/Ledger/Conformance.hs

```diff
@@ -418,3 +504,140 @@ instance
+  agdaUtxoState <- case toSpecRep $ nes ^. nesEsL . esLStateL . lsUTxOStateL of
+    Right x -> pure x
+    Left e -> fail $ T.unpack e
+  agdaTx <- case toSpecRep tx of
+    Right x -> pure x
+    Left e -> fail $ T.unpack e
+  logEntry $ "Impl Tx:\n" <> showExpr tx
+  logEntry $ "Agda Tx:\n" <> showExpr agdaTx
+  logCurPParams
+  logEntry $
+    unlines
+      [ "txins ≢ ∅\n  " <> show (not $ null txins)
+      , " txins ⊆ dom utxo\n  " <> show (txins `isSubsetOf` Map.keysSet (unUTxO utxo))
+      , " inInterval slot txvldt\n  " <> show (inInterval slot txvldt)
+      , " minfee pp tx ≤ txfee\n  " <> show (getMinFeeTx pp tx <= txfee)
+      , " consumed pp s txb ≡ produced pp s txb\n  " <> show (consumedVal == producedVal)
+      , " txsize ≤ maxTxSize pp\n  " <> show (txsize <= fromIntegral maxTxSize)
+      , " ∀[ (_ , txout) ∈ txouts .proj₁ ] inject (utxoEntrySize txout * minUTxOValue pp) ≤ᵗ getValue txout\n  "
+          <> show (all (\txout -> inject (Coin 0) <= txout ^. valueTxOutL) txouts)
+      , " ∀[ (_ , txout) ∈ txouts .proj₁ ] serSize (getValue txout) ≤ maxValSize pp\n  "
+          <> show (all (\txout -> serSize (txout ^. valueTxOutL) <= maxValSize) txouts)
+      , " ∀[ (a , _) ∈ range txouts ] Sum.All (const ⊤) (λ a → a .BootstrapAddr.attrsSize ≤ 64) a\n  "
+          -- I assume we never create bootstrap addresses
+          <> show (all (\txout -> case txout ^. addrTxOutL of AddrBootstrap _ -> False; _ -> True) txouts)
+      ]
+  logEntry $
+    unlines
+      [ "Implementation:"
+      , "  consumed:\t" ++ show (coin consumedVal)
+      , "    ins:\t\t" ++ show (coinBalance (UTxO (Map.restrictKeys (unUTxO utxo) (txb ^. inputsTxBodyL))))
+      , "  produced:\t" ++ show (coin producedVal)
+      , "    outs:\t" ++ show (coin $ sumAllValue txouts)
+      , "    fee:\t\t" ++ show txfee
+      ]
+  lastTick <- gets $ view impLastTickL
+  pParams <- getsNES $ nesEsL . curPParamsEpochStateL
+  certState <- getsNES $ nesEsL . esLStateL . lsCertStateL
+  let
+    utxoEnv =
+      UtxoEnv
+        { ueSlot = lastTick
+        , uePParams = pParams
+        , ueGenDelegs = certState ^. certDStateL . dsGenDelegsL
+        , ueCertState = certState
+        }
+  agdaUtxoEnv <- case toSpecRep utxoEnv of
+    Right x -> pure x
+    Left e -> fail $ T.unpack e
+  let
+    newAgdaUtxoState = Agda.utxoStep agdaUtxoEnv agdaUtxoState agdaTx
+  modify $ impDoFixupL .~ False
+  submitRes <- trySubmitTx tx
+  modify $ impDoFixupL .~ doFixup
+  newUtxoState <- getsNES $ nesEsL . esLStateL . lsUTxOStateL
+  let
+    finalAgdaState =
+      toExpr . specToTestRep @(UTxOState Conway) <$> newAgdaUtxoState
+    finalImplState = case submitRes of
+      Right _ -> case toTestRep newUtxoState of
+        Right x -> Just $ toExpr x
+        Left _ -> Nothing
+      Left e ->
+        error $
+          "Failed to submit transaction on the implementation side:\n"
+            <> show e
+  unless (finalAgdaState == finalImplState) $ do
```

Comment:
I guess you didn't find a better way to compare values without going through Show or ToExpr?

Reply:
I think using `Expr` for comparison works just fine, I had to add some functions to `Map` and `Set` to deal with the ordering of elements, but otherwise `Expr` gives me all I need.

Reply:
but there's an option to use orphan `Eq` instances if we want to separate prettyprinting from comparing.

Reply:
> there's an option to use orphan Eq instances if we want to separate pretty printing from comparing

That is a way better solution!

Reply:
but then we might get a case where diffing and `Eq` instance behave differently so we don't get any useful information from diff. I think it's better to make sure that the equality on `ToExpr` expressions behaves like `Eq`